src/noit_metric_tag_search.c

```diff
@@ -35,53 +35,140 @@
   char *str;
   void *impl_data;
   const noit_var_match_impl_t *impl;
 } noit_var_match_t;
 
+typedef struct re_matcher {
+  char *re_str;
+  pcre *re;
+  pcre_extra *re_e;
+  int error_offset;
+} re_matcher_t;
+
+typedef struct tls_state {
+  pcre_jit_stack *jit_stack;
+  mtev_lfu_t *pattern_to_re_matcher;
+} tls_state_t;
+
 static mtev_boolean noit_match_str(const char *subj, int subj_len, const struct noit_var_match_t *m);
+static tls_state_t *tls_state_alloc(void);
+static pcre_jit_stack *tls_state_get_pcre_jit_stack(void *arg);
+static pthread_key_t tls_state_key;
 
-static pthread_mutex_t noit_tags_search_tls_jit_init_lock = PTHREAD_MUTEX_INITIALIZER;
-static __thread pcre_jit_stack *noit_tag_search_tls_jit_stack;
-static bool noit_tags_search_tls_jit_init = false;
-static pthread_key_t noit_tag_search_tls_jit_stack_key;
+static re_matcher_t *
+re_matcher_alloc(char *pattern)
+{
+  const char *error;
+  re_matcher_t *rem = malloc(sizeof(*rem));
+
+  rem->re_str = pattern;
+  rem->re = pcre_compile(pattern, 0, &error, &rem->error_offset, NULL);
+
+#ifdef PCRE_STUDY_JIT_COMPILE
+  rem->re_e = pcre_study(rem->re, PCRE_STUDY_JIT_COMPILE, &error);
+  pcre_assign_jit_stack(rem->re_e, tls_state_get_pcre_jit_stack, NULL);
+#else
+  rem->re_e = pcre_study(rem->re, 0, &error);
+#endif
+
+  return rem;
+}
 
 static void
-noit_tag_search_free_tls_pcre_jit_stack(void *stack)
+re_matcher_free(void *v_rem)
 {
-  pcre_jit_stack_free((pcre_jit_stack *)stack);
+  re_matcher_t *rem = v_rem;
+
+  if (rem) {
+    if (rem->re_e) {
+#ifdef PCRE_STUDY_JIT_COMPILE
+      pcre_free_study(rem->re_e);
+#else
+      pcre_free(rem->re_e);
+#endif
+    }
+
+    pcre_free(rem->re);
+    free(rem->re_str);
+    free(rem);
+  }
 }
-static pcre_jit_stack *
-noit_tag_search_get_tls_pcre_jit_stack(void *arg)
+
+static void
+tls_state_free(void *v)
+{
+  tls_state_t *st = v;
+
+  pcre_jit_stack_free(st->jit_stack);
+  mtev_lfu_destroy(st->pattern_to_re_matcher);
+  free(st);
+}
+
+static tls_state_t *
+tls_state_alloc(void)
+{
+  tls_state_t *st = pthread_getspecific(tls_state_key);
+
+  if (st == NULL) {
+    st = malloc(sizeof(tls_state_t));
+    st->jit_stack = pcre_jit_stack_alloc(1024 * 32, 512 * 1024);
+    st->pattern_to_re_matcher = mtev_lfu_create(100000, re_matcher_free);
+    pthread_setspecific(tls_state_key, st);
+  }
+
+  return st;
+}
+
+static pcre_jit_stack *tls_state_get_pcre_jit_stack(void *arg)
 {
   (void)arg;
-  if (noit_tags_search_tls_jit_init && noit_tag_search_tls_jit_stack == NULL) {
-    noit_tag_search_tls_jit_stack = (pcre_jit_stack *)pthread_getspecific(noit_tag_search_tls_jit_stack_key);
-    if (noit_tag_search_tls_jit_stack == NULL) {
-      noit_tag_search_tls_jit_stack = (pcre_jit_stack *)pcre_jit_stack_alloc(1024 * 32, 512 * 1024);
-      pthread_setspecific(noit_tag_search_tls_jit_stack_key, noit_tag_search_tls_jit_stack);
-    }
+  return tls_state_alloc()->jit_stack;
+}
+
+static re_matcher_t *
+tls_state_get_pattern(char *pattern)
+{
```

Comment:
Send like pattern is allocated, not feed by caller or here and leaks

Reply:
it's stored in the LFU and freed in ``re_matcher_free``